axlearn/common/loss.py

```diff
@@ -141,64 +146,80 @@ def cross_entropy(
         targets = _one_hot_with_label_smoothing(
             target_labels, num_classes, label_smoothing=label_smoothing
         )
-    pre_mask_loss, pre_mask_cross_entropy_loss, pre_mask_z_loss = _stable_cross_entropy(
+    per_target_loss, per_target_cross_entropy_loss, per_target_z_loss = _stable_cross_entropy(
         logits, targets, z_loss_scale
     )
-    if mask is None:
-        mask = jnp.logical_and(0 <= target_labels, target_labels < num_classes)
-    mask = mask.astype(pre_mask_loss.dtype)
-    num_unmasked = jnp.maximum(mask.sum(), 1)
-    cross_entropy_loss = (pre_mask_cross_entropy_loss * mask).sum() / num_unmasked
-    z_loss = (pre_mask_z_loss * mask).sum() / num_unmasked
-    loss = (pre_mask_loss * mask).sum() / num_unmasked
+    if mask is not None:
+        if live_targets is None:
+            live_targets = mask
+        else:
+            raise ValueError("mask and live_targets must not be specified together")
+    if live_targets is None:
+        live_targets = jnp.logical_and(0 <= target_labels, target_labels < num_classes)
+    live_targets = live_targets.astype(per_target_loss.dtype)
+    denominator = jnp.maximum(live_targets.sum(), 1)
+    cross_entropy_loss = (per_target_cross_entropy_loss * live_targets).sum() / denominator
+    z_loss = (per_target_z_loss * live_targets).sum() / denominator
+    loss = (per_target_loss * live_targets).sum() / denominator
     predicted_labels = jnp.argmax(logits, axis=-1)
-    accuracy = (jnp.equal(predicted_labels, target_labels) * mask).sum() / num_unmasked
+    accuracy = (jnp.equal(predicted_labels, target_labels) * live_targets).sum() / denominator
```

Comment:
Hi Ruoming, should we use this accuracy inside CrossEntropyLossMetrics::forward() instead of recalculating it in CrossEntropyLossMetrics::forward() ?
see: https://github.com/apple/axlearn/blob/main/axlearn/common/causal_lm.py#L108C1-L117C10
@ruomingp